### ISSUE TITLE
AD nitpick

### ISSUE THREAD

**Issue by MarisaKirisame** [GiggleLiu/nilangpaper#5]
> Instruction level automatic differentiation on a reversible Turing machine

This has nothing to do with a reversible Turing machine. Instead, it was a reversible Programming Language.
The major difference is that nilang is built on a modern computer which does random access, while a turing machine only has linearized access pattern, making nonlocal dataflow slow.
You are also transforming the code from nilang into niliang, instead of first compiling nilang to an intermediate language. This way higher level abstraction is retained - for example, the algorithm is now able to do transformtion on reverse while loop, or use precondition to help transform (which I think is impossible on a reversible turing machine).

> For source to source AD package, Tapenade~\cite{Hascoet2013} uses source code as the computational graph and Zygote~\cite{Innes2018, Innes2019} uses an intermediate representation (IR) of a program, the static single assignment (SSA) form, as the computational graph. To cache intermediate states, they use a global stack.

The computation graph (in AD) by definition is linear in nature - there is no any kind of control flow. tapenede, zygote all have an intermediate IR and do source to source transform on it. The computation graph does not manifest at runtime in the either approach (but can be recovered from the wengert list).

> First of all, most packages require a lot of primitive functions with programmer-defined backward rules. For example, the backward rule of Bessel functions must be provided although it is composed of basic instructions `+', `-', `*', `/', and conditional jumps.

See evaluating deriative ~section 2.2. Tapanede can technically do adjoint over arbitary floating point function, but it is heavily advised to not do so because of floating point instability and computational efficiency.

> Defining backward rules for these basic instructions in the computational graph scheme suffers from the overhead of memorizing the computational graph and caching intermediate states. Even in Tapenade, the program has to remember the control at each place where the flow merges in the forward sweep.

This point is orthgonal to the fact that they are basic instruction. This is just an reiteration of point 2 (The overhead of naively caching...).

> Another important source of memory overhead is from the fact that inplace functions are forbidden delibrately in a computational graph based AD scheme in order to protect the cached data.

inplace functions are allowed both in traditional and deep learning based ad framework.
in classical ad framework, values can be inplace modify, but the old version will be saved on the list, if needed for gradient/unable to recovered from the future value (aka adjointing the linear evolution) in MXNet (see tq's checkpointing paper), liveness analysis is applied to turn operation into their in-place counterpart. It is true that one cannot really use inplace to save memory/compute unless the old value is not needed, so a better way of saying this is "inplace cannot be used to save memory because both the original value and the new value is needed"

> Thirdly, obtaining higher-order gradients are not efficient in most of these packages. For example, in most machine learning packages, people back-propagate the whole program of obtaining first-order gradients to obtain second-order gradients. The repeated use of back-propagation algorithm causes an exponential overhead concerning the order of gradients. A better approach is using Taylor propagation like in JAX~\cite{Bettencourt2019} and beautiful differentiation~\cite{Elliott2009}. However, Taylor propagation in the adjoint mode AD requires tedious implemention of higher order backward rules for primitives.

This is inaccurate. In pytorch, people backprop once to obtain a computation graph containing only the gradient path, then they can backprop only that graph. In classical AD, a To-Be-Record (TBR) analysis will be done to figure out that the original code does not need first order AD again.

I feel like the first point and the third point should be removed, focusing the story on the second point - saving more memory.

> With respect to the order of gradients, Taylor propagation is exponentially more efficient in obtaining higher-order gradients than differentiating lower order gradients recursively.

Please remove exponential for reason discuss above.

> EVAL

I see a "split brain" problem with the eval.
OTOH, the story talk about the memory wall and use deep learning as an example, while on the other hand, all of the eval is focused on non deep learning task. I think we should get more invertible network in the eval.

Finally, can we cite the TREEVERSE paper? It is a very famous paper in checkpointing. (I think a related work section should be written as well).

This is a very rough pass, in particular I havent checked claims about higher order AD as I am not an expert in them. I will take a look sometimes later.

**Comment by GiggleLiu** [comment]
Agree with most of these changes, especially the comments to higher order AD is not so correct. do you mind modefying them directly in our paper? You may have better understanding about the problem of "can not differentiate a general program".
A question

>  transforming the code from nilang into nilang

Do you mean preprocessing? Why is it related to Turing machine?

## About workflow,
* We can use the script `.tdiff` to see the changes. e.g. `./tdiff 3` will show the change comparing with HEAD~3.
* You can still comment in our paper using colored texts. See the example in the first line of the introduction.

**Comment by MarisaKirisame** [comment]
I could change the paper. I just want to let you know what I'm about to change as this is a heavy modification of the current version.

I mean the adjoint AD algorithm. It goes from an structural IR to a structural IR, not turing machine to turing machine.

**Comment by GiggleLiu** [comment]
ok, I see.
